### PR TITLE
MTDSA-1071 - Return 500 when lookup service receives unhandled respon…

### PR DESCRIPTION
**Pull request by hicnar**
…se from DES.

### CODE REVIEW

File: app/uk/gov/hmrc/selfassessmentapi/services/AuthenticationService.scala [hunk 1]
```diff
@@ -43,8 +48,12 @@ object AuthenticationService extends AuthorisedFunctions {
 
   def authCheck(nino: Nino)(implicit hc: HeaderCarrier, reqHeader: RequestHeader): Future[AuthResult] =
     lookupService.mtdReferenceFor(nino).flatMap {
-      case Some(id) => authoriseAsClient(id)
-      case None => Future.successful(Left(Forbidden(Json.toJson(Errors.ClientNotSubscribed))))
+      case Right(id) => authoriseAsClient(id)
+      case Left(status) => status match {
+        case 400 =>  Future.successful(Left(Status(BAD_REQUEST)(toJson(NinoInvalid))))
```
Comment (GlassAndOneHalf) [inline app/uk/gov/hmrc/selfassessmentapi/services/AuthenticationService.scala]: `Left(BadRequest(toJson(NinoInvalid))` is a bit easier to read.

File: app/uk/gov/hmrc/selfassessmentapi/services/MtdRefLookupService.scala [hunk 2]
```diff
@@ -31,32 +31,40 @@ trait MtdRefLookupService {
...
             repository.store(nino, id)
             id
-          }
-        case 400 | 404 =>
-          logger.debug(s"NINO to MTD reference lookup failed. Status code: [${response.status}]")
-          None
-        case _ =>
-          logger.error(s"NINO to MTD reference lookup failed. Is there an issue with DES? Status code: [${response.status}]")
-          None
+          }.getOrElse(throw new IllegalStateException("Unable to retrieve the mtdId")))
```
Comment (dabd) [inline app/uk/gov/hmrc/selfassessmentapi/services/MtdRefLookupService.scala]: Can we encode this persistence error as a value and not throw the exception?  You could define a sealed trait `Error`, two subtypes `PersistenceError`, `HttpError`, and make `cacheReferenceFor` return `Future[Either[Error, MtdId]]`, for example.

File: app/uk/gov/hmrc/selfassessmentapi/services/AuthenticationService.scala [hunk 3]
```diff
@@ -83,7 +92,7 @@ object AuthenticationService extends AuthorisedFunctions {
       Future.successful(Left(Forbidden(Json.toJson(Errors.AgentNotSubscribed))))
     case _: UnsupportedAffinityGroup =>
       logger.debug(s"Authorisation failed as client.")
-      Future.successful(Left(Forbidden(Json.toJson(Errors.ClientNotSubscribed))))
+      Future.successful(Left(Forbidden(Json.toJson(ClientNotSubscribed))))
```
Comment (GlassAndOneHalf) [inline app/uk/gov/hmrc/selfassessmentapi/services/AuthenticationService.scala]: `Json.toJson` could be `toJson`

**Comment by dabd** [comment]
👍

**Pull request merged**